// prettier-plugin-apex/dist/bin/start-apex-server.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const yargs_1 = __importDefault(require("yargs"));
const helpers_1 = require("yargs/helpers");
const http_server_1 = require("../src/http-server");
async function setup(host, port, allowedOrigins) {
    await (0, http_server_1.start)(host, port, allowedOrigins);
}
(0, yargs_1.default)((0, helpers_1.hideBin)(process.argv))
    .command("$0", "start the built-in parsing server", {
    host: {
        alias: "h",
        default: "localhost",
    },
    port: {
        alias: "p",
        default: 2117,
    },
    "cors-allowed-origins": {
        describe: "Comma-delimited list of allowed origins to be added to CORS headers",
        alias: "c",
        type: "string",
    },
}, (argv) => {
    setup(argv.host, argv.port, argv["cors-allowed-origins"]);
})
    .help()
    .parse();
