// @lwc/template-compiler/dist/commonjs/shared/ast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isIfBlock = exports.isIf = exports.isForBlock = exports.isForEach = exports.isForOf = exports.isBooleanLiteral = exports.isStringLiteral = exports.isExpression = exports.isComment = exports.isText = exports.isLwcComponent = exports.isBaseLwcElement = exports.isBaseElement = exports.isSlot = exports.isComponent = exports.isExternalComponent = exports.isRoot = exports.isElement = exports.property = exports.attribute = exports.renderModeDirective = exports.preserveCommentsDirective = exports.refDirective = exports.innerHTMLDirective = exports.domDirective = exports.slotDataDirective = exports.slotBindDirective = exports.spreadDirective = exports.lwcIsDirective = exports.dynamicDirective = exports.keyDirective = exports.eventListener = exports.elseBlockNode = exports.elseifBlockNode = exports.ifBlockNode = exports.ifNode = exports.scopedSlotFragment = exports.forOf = exports.forEach = exports.literal = exports.sourceLocation = exports.elementSourceLocation = exports.comment = exports.text = exports.slot = exports.lwcComponent = exports.component = exports.externalComponent = exports.element = exports.root = void 0;
exports.isScopedSlotFragment = exports.isProperty = exports.isPreserveCommentsDirective = exports.isRenderModeDirective = exports.isSlotBindDirective = exports.isSlotDataDirective = exports.isKeyDirective = exports.isRefDirective = exports.isInnerHTMLDirective = exports.isSpreadDirective = exports.isDomDirective = exports.isLwcIsDirective = exports.isDynamicDirective = exports.isParentNode = exports.isElementDirective = exports.isConditionalBlock = exports.isConditionalParentBlock = exports.isElseBlock = exports.isElseifBlock = void 0;
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const shared_1 = require("@lwc/shared");
function root(parse5ElmLocation) {
    return {
        type: 'Root',
        location: elementSourceLocation(parse5ElmLocation),
        directives: [],
        children: [],
    };
}
exports.root = root;
function element(tagName, namespaceURI, parse5ElmLocation) {
    return {
        type: 'Element',
        name: tagName,
        namespace: namespaceURI,
        location: elementSourceLocation(parse5ElmLocation),
        attributes: [],
        properties: [],
        directives: [],
        listeners: [],
        children: [],
    };
}
exports.element = element;
function externalComponent(tagName, parse5ElmLocation) {
    return {
        type: 'ExternalComponent',
        name: tagName,
        namespace: shared_1.HTML_NAMESPACE,
        location: elementSourceLocation(parse5ElmLocation),
        attributes: [],
        properties: [],
        directives: [],
        listeners: [],
        children: [],
    };
}
exports.externalComponent = externalComponent;
function component(tagName, parse5ElmLocation) {
    return {
        type: 'Component',
        name: tagName,
        namespace: shared_1.HTML_NAMESPACE,
        location: elementSourceLocation(parse5ElmLocation),
        attributes: [],
        properties: [],
        directives: [],
        listeners: [],
        children: [],
    };
}
exports.component = component;
function lwcComponent(tagName, parse5ElmLocation) {
    return {
        type: 'Lwc',
        name: tagName,
        namespace: shared_1.HTML_NAMESPACE,
        location: elementSourceLocation(parse5ElmLocation),
        attributes: [],
        properties: [],
        directives: [],
        listeners: [],
        children: [],
    };
}
exports.lwcComponent = lwcComponent;
function slot(slotName, parse5ElmLocation) {
    return {
        type: 'Slot',
        name: 'slot',
        namespace: shared_1.HTML_NAMESPACE,
        slotName,
        location: elementSourceLocation(parse5ElmLocation),
        attributes: [],
        properties: [],
        directives: [],
        listeners: [],
        children: [],
    };
}
exports.slot = slot;
function text(raw, 
// TODO [#3370]: remove experimental template expression flag
value, parse5Location) {
    return {
        type: 'Text',
        raw,
        value,
        location: sourceLocation(parse5Location),
    };
}
exports.text = text;
function comment(raw, value, parse5Location) {
    return {
        type: 'Comment',
        raw,
        value,
        location: sourceLocation(parse5Location),
    };
}
exports.comment = comment;
function elementSourceLocation(parse5ElmLocation) {
    const elementLocation = sourceLocation(parse5ElmLocation);
    const startTag = sourceLocation(parse5ElmLocation.startTag);
    // endTag must be optional because Parse5 currently fails to collect end tag location for element with a tag name
    // containing an upper case character (inikulin/parse5#352).
    const endTag = parse5ElmLocation.endTag
        ? sourceLocation(parse5ElmLocation.endTag)
        : parse5ElmLocation.endTag;
    return { ...elementLocation, startTag, endTag };
}
exports.elementSourceLocation = elementSourceLocation;
function sourceLocation(location) {
    return {
        startLine: location.startLine,
        startColumn: location.startCol,
        endLine: location.endLine,
        endColumn: location.endCol,
        start: location.startOffset,
        end: location.endOffset,
    };
}
exports.sourceLocation = sourceLocation;
function literal(value) {
    return {
        type: 'Literal',
        value,
    };
}
exports.literal = literal;
function forEach(expression, elementLocation, directiveLocation, item, index) {
    return {
        type: 'ForEach',
        expression,
        item,
        index,
        location: elementLocation,
        directiveLocation,
        children: [],
    };
}
exports.forEach = forEach;
function forOf(expression, iterator, elementLocation, directiveLocation) {
    return {
        type: 'ForOf',
        expression,
        iterator,
        location: elementLocation,
        directiveLocation,
        children: [],
    };
}
exports.forOf = forOf;
function scopedSlotFragment(identifier, elementLocation, directiveLocation, slotName) {
    return {
        type: 'ScopedSlotFragment',
        location: elementLocation,
        directiveLocation,
        children: [],
        slotData: slotDataDirective(identifier, directiveLocation),
        slotName: slotName,
    };
}
exports.scopedSlotFragment = scopedSlotFragment;
function ifNode(modifier, condition, elementLocation, directiveLocation) {
    return {
        type: 'If',
        modifier,
        condition,
        location: elementLocation,
        directiveLocation,
        children: [],
    };
}
exports.ifNode = ifNode;
function ifBlockNode(condition, elementLocation, directiveLocation) {
    return {
        type: 'IfBlock',
        condition,
        location: elementLocation,
        directiveLocation,
        children: [],
    };
}
exports.ifBlockNode = ifBlockNode;
function elseifBlockNode(condition, elementLocation, directiveLocation) {
    return {
        type: 'ElseifBlock',
        condition,
        location: elementLocation,
        directiveLocation,
        children: [],
    };
}
exports.elseifBlockNode = elseifBlockNode;
function elseBlockNode(elementLocation, directiveLocation) {
    return {
        type: 'ElseBlock',
        location: elementLocation,
        directiveLocation,
        children: [],
    };
}
exports.elseBlockNode = elseBlockNode;
function eventListener(name, handler, location) {
    return {
        type: 'EventListener',
        name,
        handler,
        location,
    };
}
exports.eventListener = eventListener;
function keyDirective(value, location) {
    return {
        type: 'Directive',
        name: 'Key',
        value,
        location,
    };
}
exports.keyDirective = keyDirective;
function dynamicDirective(value, location) {
    return {
        type: 'Directive',
        name: 'Dynamic',
        value,
        location,
    };
}
exports.dynamicDirective = dynamicDirective;
function lwcIsDirective(value, location) {
    return {
        type: 'Directive',
        name: 'Is',
        value,
        location,
    };
}
exports.lwcIsDirective = lwcIsDirective;
function spreadDirective(value, location) {
    return {
        type: 'Directive',
        name: 'Spread',
        value,
        location,
    };
}
exports.spreadDirective = spreadDirective;
function slotBindDirective(value, location) {
    return {
        type: 'Directive',
        name: 'SlotBind',
        value,
        location,
    };
}
exports.slotBindDirective = slotBindDirective;
function slotDataDirective(value, location) {
    return {
        type: 'Directive',
        name: 'SlotData',
        value,
        location,
    };
}
exports.slotDataDirective = slotDataDirective;
function domDirective(lwcDomAttr, location) {
    return {
        type: 'Directive',
        name: 'Dom',
        value: literal(lwcDomAttr),
        location,
    };
}
exports.domDirective = domDirective;
function innerHTMLDirective(value, location) {
    return {
        type: 'Directive',
        name: 'InnerHTML',
        value,
        location,
    };
}
exports.innerHTMLDirective = innerHTMLDirective;
function refDirective(value, location) {
    return {
        type: 'Directive',
        name: 'Ref',
        value,
        location,
    };
}
exports.refDirective = refDirective;
function preserveCommentsDirective(preserveComments, location) {
    return {
        type: 'Directive',
        name: 'PreserveComments',
        value: literal(preserveComments),
        location,
    };
}
exports.preserveCommentsDirective = preserveCommentsDirective;
function renderModeDirective(renderMode, location) {
    return {
        type: 'Directive',
        name: 'RenderMode',
        value: literal(renderMode),
        location,
    };
}
exports.renderModeDirective = renderModeDirective;
function attribute(name, value, location) {
    return {
        type: 'Attribute',
        name,
        value,
        location,
    };
}
exports.attribute = attribute;
function property(name, attributeName, value, location) {
    return {
        type: 'Property',
        name,
        attributeName,
        value,
        location,
    };
}
exports.property = property;
function isElement(node) {
    return node.type === 'Element';
}
exports.isElement = isElement;
function isRoot(node) {
    return node.type === 'Root';
}
exports.isRoot = isRoot;
function isExternalComponent(node) {
    return node.type === 'ExternalComponent';
}
exports.isExternalComponent = isExternalComponent;
function isComponent(node) {
    return node.type === 'Component';
}
exports.isComponent = isComponent;
function isSlot(node) {
    return node.type === 'Slot';
}
exports.isSlot = isSlot;
function isBaseElement(node) {
    return (isElement(node) ||
        isComponent(node) ||
        isSlot(node) ||
        isExternalComponent(node) ||
        isLwcComponent(node));
}
exports.isBaseElement = isBaseElement;
// BaseLwcElement represents special LWC tags denoted lwc:*
function isBaseLwcElement(node) {
    return node.type === 'Lwc';
}
exports.isBaseLwcElement = isBaseLwcElement;
// Represents the lwc:component tag
function isLwcComponent(node) {
    return isBaseLwcElement(node) && node.name === 'lwc:component';
}
exports.isLwcComponent = isLwcComponent;
function isText(node) {
    return node.type === 'Text';
}
exports.isText = isText;
function isComment(node) {
    return node.type === 'Comment';
}
exports.isComment = isComment;
function isExpression(node) {
    return node.type !== 'Literal';
}
exports.isExpression = isExpression;
function isStringLiteral(node) {
    return node.type === 'Literal' && typeof node.value === 'string';
}
exports.isStringLiteral = isStringLiteral;
function isBooleanLiteral(node) {
    return node.type === 'Literal' && typeof node.value === 'boolean';
}
exports.isBooleanLiteral = isBooleanLiteral;
function isForOf(node) {
    return node.type === 'ForOf';
}
exports.isForOf = isForOf;
function isForEach(node) {
    return node.type === 'ForEach';
}
exports.isForEach = isForEach;
function isForBlock(node) {
    return isForOf(node) || isForEach(node);
}
exports.isForBlock = isForBlock;
function isIf(node) {
    return node.type === 'If';
}
exports.isIf = isIf;
function isIfBlock(node) {
    return node.type === 'IfBlock';
}
exports.isIfBlock = isIfBlock;
function isElseifBlock(node) {
    return node.type === 'ElseifBlock';
}
exports.isElseifBlock = isElseifBlock;
function isElseBlock(node) {
    return node.type === 'ElseBlock';
}
exports.isElseBlock = isElseBlock;
function isConditionalParentBlock(node) {
    return isIfBlock(node) || isElseifBlock(node);
}
exports.isConditionalParentBlock = isConditionalParentBlock;
function isConditionalBlock(node) {
    return isIfBlock(node) || isElseifBlock(node) || isElseBlock(node);
}
exports.isConditionalBlock = isConditionalBlock;
function isElementDirective(node) {
    return isConditionalBlock(node) || isForBlock(node) || isIf(node) || isScopedSlotFragment(node);
}
exports.isElementDirective = isElementDirective;
function isParentNode(node) {
    return isBaseElement(node) || isRoot(node) || isForBlock(node) || isIf(node);
}
exports.isParentNode = isParentNode;
function isDynamicDirective(directive) {
    return directive.name === 'Dynamic';
}
exports.isDynamicDirective = isDynamicDirective;
function isLwcIsDirective(directive) {
    return directive.name === 'Is';
}
exports.isLwcIsDirective = isLwcIsDirective;
function isDomDirective(directive) {
    return directive.name === 'Dom';
}
exports.isDomDirective = isDomDirective;
function isSpreadDirective(directive) {
    return directive.name === 'Spread';
}
exports.isSpreadDirective = isSpreadDirective;
function isInnerHTMLDirective(directive) {
    return directive.name === 'InnerHTML';
}
exports.isInnerHTMLDirective = isInnerHTMLDirective;
function isRefDirective(directive) {
    return directive.name === 'Ref';
}
exports.isRefDirective = isRefDirective;
function isKeyDirective(directive) {
    return directive.name === 'Key';
}
exports.isKeyDirective = isKeyDirective;
function isSlotDataDirective(directive) {
    return directive.name === 'SlotData';
}
exports.isSlotDataDirective = isSlotDataDirective;
function isSlotBindDirective(directive) {
    return directive.name === 'SlotBind';
}
exports.isSlotBindDirective = isSlotBindDirective;
function isRenderModeDirective(directive) {
    return directive.name === 'RenderMode';
}
exports.isRenderModeDirective = isRenderModeDirective;
function isPreserveCommentsDirective(directive) {
    return directive.name === 'PreserveComments';
}
exports.isPreserveCommentsDirective = isPreserveCommentsDirective;
function isProperty(node) {
    return node.type === 'Property';
}
exports.isProperty = isProperty;
function isScopedSlotFragment(node) {
    return node.type === 'ScopedSlotFragment';
}
exports.isScopedSlotFragment = isScopedSlotFragment;
