// @lwc/template-compiler/dist/commonjs/shared/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RootDirectiveName = exports.ElementDirectiveName = exports.TemplateDirectiveName = exports.LwcTagName = exports.LWCDirectiveRenderMode = exports.LWCDirectiveDomMode = void 0;
var LWCDirectiveDomMode;
(function (LWCDirectiveDomMode) {
    LWCDirectiveDomMode["manual"] = "manual";
})(LWCDirectiveDomMode = exports.LWCDirectiveDomMode || (exports.LWCDirectiveDomMode = {}));
var LWCDirectiveRenderMode;
(function (LWCDirectiveRenderMode) {
    LWCDirectiveRenderMode["shadow"] = "shadow";
    LWCDirectiveRenderMode["light"] = "light";
})(LWCDirectiveRenderMode = exports.LWCDirectiveRenderMode || (exports.LWCDirectiveRenderMode = {}));
/**
 * All supported special LWC tags, they should all begin with lwc:*
 */
var LwcTagName;
(function (LwcTagName) {
    LwcTagName["Component"] = "lwc:component";
})(LwcTagName = exports.LwcTagName || (exports.LwcTagName = {}));
var TemplateDirectiveName;
(function (TemplateDirectiveName) {
    TemplateDirectiveName["If"] = "if:true";
    TemplateDirectiveName["IfBlock"] = "lwc:if";
    TemplateDirectiveName["ElseifBlock"] = "lwc:elseif";
    TemplateDirectiveName["ElseBlock"] = "lwc:else";
    TemplateDirectiveName["ForEach"] = "for:each";
    TemplateDirectiveName["ForOf"] = "for:of";
    TemplateDirectiveName["ScopedSlotFragment"] = "lwc:slot-data";
})(TemplateDirectiveName = exports.TemplateDirectiveName || (exports.TemplateDirectiveName = {}));
var ElementDirectiveName;
(function (ElementDirectiveName) {
    ElementDirectiveName["Dom"] = "lwc:dom";
    // TODO [#3331]: remove usage of lwc:dynamic in 246
    ElementDirectiveName["Dynamic"] = "lwc:dynamic";
    ElementDirectiveName["Is"] = "lwc:is";
    ElementDirectiveName["External"] = "lwc:external";
    ElementDirectiveName["InnerHTML"] = "lwc:inner-html";
    ElementDirectiveName["Ref"] = "lwc:ref";
    ElementDirectiveName["SlotBind"] = "lwc:slot-bind";
    ElementDirectiveName["SlotData"] = "lwc:slot-data";
    ElementDirectiveName["Spread"] = "lwc:spread";
    ElementDirectiveName["Key"] = "key";
})(ElementDirectiveName = exports.ElementDirectiveName || (exports.ElementDirectiveName = {}));
var RootDirectiveName;
(function (RootDirectiveName) {
    RootDirectiveName["PreserveComments"] = "lwc:preserve-comments";
    RootDirectiveName["RenderMode"] = "lwc:render-mode";
})(RootDirectiveName = exports.RootDirectiveName || (exports.RootDirectiveName = {}));
